### PR TITLE
Add documentation on spec documentation

### PR DESCRIPTION
**Pull request by swallez**
Adds a doc explaining how to write and structure JSDoc documentation in the specification.

@goodroot I added you as a reviewer, feel free to add more people from the docs team to raise awareness and provide feedback.

### CODE REVIEW

File: docs/doc-comments-guide.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+### Terseness
+
+**Doc comments are reference material**: they should be as succinct as possible while capturing all the necessary information to use the elements they're documenting. Remember that they will often show up in small IDE autocompletion popups!
+
+In particular, doc comments are not the right place for tutorials or examples, which should be in dedicated documentation pages. These pages can of course be linked from the doc comments.
+
+API endpoints will also have a `@doc_url` JSDoc tag that links to that API's detailed documentation page.
+
+### Multi-paragraph doc comments
```
Comment (philkra) [inline docs/doc-comments-guide.md:86]: Can we merge the content with [this section](https://github.com/elastic/elasticsearch-specification/blob/main/docs/modeling-guide.md#description)?

**Review by goodroot — COMMENTED** [review]
Thanks @swallez 👍 

Appreciate the documentation. The writing advice re: how to structure annotations to be succinct, contain  a descriptive line before substantial paragraphs, is also appreciated. It is a step forward to a standard we can adopt across many contexts when generating API docs. 

Pinging writers from Elasticsearch: @debadair @jrodewig @lockewritesdocs 

And writers with a general interest in API docs: @KOTungseth @lcawl @kilfoyle 

Folks: this PR documents how the API specifications now support GitHub Flavoured Markdown. These annotated TypeScript API specifications render down into JSON which an eventual "next docs" generator will turn into MDX doc pages.  Happy to unpack in greater detail and no action is required, but feel free to leave feedback.

**Review by delvedor — COMMENTED** [review]
I'm wondering if we should also add a check in the compiler to verify if the `description` can be parsed with GFM.

File: docs/doc-comments-guide.md (hunk 1 above)
Comment (swallez) [reply docs/doc-comments-guide.md:86]: Added a [commit](https://github.com/elastic/elasticsearch-specification/pull/638/commits/8483d278157c687fc00f1224237f5d83bd9cbd72) that that adds a link to this document in that section.

File: docs/doc-comments-guide.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
+# Documenting the API specification
+
+A specification is not only about formalizing data structures, it's also about explaining what these structures and their properties mean are used for.
```
Comment (gtback) [inline docs/doc-comments-guide.md:3]:
```suggestion
A specification is not only about formalizing data structures, it's also about explaining what these structures and their properties are used for.
```

I think maybe an extra word here?

**Review by gtback — APPROVED** [review]
❤️ documentation about how to write documentation ✍🏻 .

File: docs/doc-comments-guide.md (hunk 2 above)
Comment (swallez) [reply docs/doc-comments-guide.md]: Indeed, good catch!

**Pull request merged**